src/zenml/new/pipelines/pipeline.py

```diff
@@ -0,0 +1,1088 @@
+        return invocation_id
+
+    def _compute_invocation_id(
+        self,
+        step: "BaseStep",
+        custom_id: Optional[str] = None,
+        allow_suffix: bool = True,
+    ) -> str:
+        """Compute the invocation ID.
+
+        Args:
+            step: The step for which to compute the ID.
+            custom_id: Custom ID to use for the invocation.
+            allow_suffix: Whether a suffix can be appended to the invocation
+                ID.
+
+        Raises:
+            RuntimeError: If no ID suffix is allowed and an invocation for the
+                same ID already exists.
+            RuntimeError: If no unique invocation ID can be found.
+
+        Returns:
+            The invocation ID.
+        """
+        base_id = custom_id or step.name
+
+        if base_id in self.invocations and not allow_suffix:
+            raise RuntimeError("Duplicate step ID")
+
+        id_ = base_id
+        for index in range(2, 10000):
+            if id_ not in self.invocations:
+                break
+
+            id_ = f"{base_id}_{index}"
+        else:
+            raise RuntimeError("Unable to find step ID")
+
+        return id_
+
+    def __enter__(self: T) -> T:
+        """Enter the pipeline context.
+
+        Args:
+            self: _description_
+
+        Raises:
+            RuntimeError: _description_
+
+        Returns:
+            _description_
+        """
+        if Pipeline.ACTIVE_PIPELINE:
+            raise RuntimeError("Pipeline already active.")
+
+        Pipeline.ACTIVE_PIPELINE = self
+        return self
+
+    def __exit__(self, type, value, traceback):
+        Pipeline.ACTIVE_PIPELINE = None
+
+    def with_options(
+        self,
+        run_name: Optional[str] = None,
+        schedule: Optional[Schedule] = None,
+        build: Union[str, "UUID", "PipelineBuildBaseModel", None] = None,
+        step_configurations: Optional[
+            Mapping[str, "StepConfigurationUpdateOrDict"]
+        ] = None,
+        config_path: Optional[str] = None,
+        unlisted: bool = False,
+        prevent_build_reuse: bool = False,
+        **kwargs,
+    ) -> "Pipeline":
+        pipeline_copy = self.copy()
+        pipeline_copy.configure(**kwargs)
+
+        run_args = dict_utils.remove_none_values(
+            {
+                "run_name": run_name,
+                "schedule": schedule,
+                "build": build,
+                "step_configurations": step_configurations,
+                "config_path": config_path,
+                "unlisted": unlisted,
+                "prevent_build_reuse": prevent_build_reuse,
+            }
+        )
+        pipeline_copy._run_args.update(run_args)
+        return pipeline_copy
+
+    def copy(self) -> "Pipeline":
+        """Copies the pipeline.
+
+        Returns:
+            The pipeline copy.
+        """
+        return copy.deepcopy(self)
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
```

Comment:
This needs a lot of explanations, I have no clue what's happening in this entire method